### ISSUE TITLE
`base mixin` that implements a `sealed class` counts towards exhaustiveness. Maybe allow `sealed mixin`.

### ISSUE THREAD

**Issue by ds84182** [dart-lang/language#3103]
Seeing this on `3.1.0-109.0.dev`.

Only in Analyzer: Uncommenting Foo, like the enums, complains that the case will never match.

```dart
sealed class Sealed {}

enum Enum implements Sealed {
  value;
}

enum Enum2 implements Sealed {
  value;
}

final class Subclass extends Sealed {}

base mixin Foo implements Sealed {}

enum EnumFoo with Foo {
  value;
}

// The type 'Sealed' is not exhaustively matched by the switch cases since it doesn't match 'Foo()'.
//                   vvvvvv
int foo(Sealed s) => switch (s) {
      // The matched value type 'Sealed' can never match the required type 'EnumFoo'. Try using a different pattern.
      //vv
      Enum _ => 0,
      Enum2.value => 0,
      Subclass _ => 0,
      // The matched value type 'Sealed' can never match the required type 'EnumFoo'. Try using a different pattern.
      //vvvvv
      EnumFoo _ => 0,
      // The matched value type 'Sealed' can never match the required type 'Foo'. Try using a different pattern.
      // vvv
      // Foo _ => 0,
    };

void main() {
  foo(Enum.value);
}
```

**Comment by mateusfccp** [comment]
I agree that `sealed mixin` should be a thing. It should only prevent mixing outside of library, like `sealed class` do with extension. I don't think they are mutually exclusive and incompatible.

**Comment by lrhn** [comment]
Another case of this:  https://github.com/dart-lang/sdk/issues/53809

**Comment by TekExplorer** [comment]
> it is considered too confusing to have something visibly marked `mixin` which cannot be mixed in.

That doesnt make sense to me. if its `sealed`, which is the first thing you see, then you already know it cant be used at all by your classes, so it being a `mixin` or not is completely irrelevant, and should be considered an implementation detail.

that implies that `sealed class` is confusing because its a `class` you cant extend. perhaps its not exactly the same, but its similar enough.

I often find myself wanting to use `sealed mixin`s so that I can take advantage of `mixin` features while achieving exhaustiveness.

I see no reason why `sealed` would not be allowed on _any_ implementable type, `mixin`, `class`, or any combination thereof.

hell, perhaps the analyzer should just not show anything other than `sealed` in hints, because thats all consumers should care about.

**Comment by eernstg** [comment]
An example illustrating the need for sealed mixins came up here: https://github.com/dart-lang/sdk/issues/63313.

**Comment by lrhn** [comment]
Dart has locked down both of the ways a mixin can be used with a sealed hierarachy:
* `sealed mixin Foo` as the root of the sealed hierarchy
* `mixin Foo on SealedSuper ` being allowed, and cannot be implemented, without counting towards exhaustion of `SealedSuper`. (Because all concrete subclasses must apply it to one of the subclasses which do count.)

Without either of these, mixins simply don't jive with sealed.

**Comment by eernstg** [comment]
@lrhn, I don't know what you mean by 'Dart has locked down', but the example [here](https://github.com/dart-lang/sdk/issues/63313#issuecomment-4379955279) shows that sealed mixins jive just fine as nodes in a superinterface graph containing a number of other sealed declarations: The classes `ForElement`, `Expression`, `SpreadElement`, `IfElement`, `MapLiteralEntry`, and `NullAwareElement` is an exhaustive set of types for `CollectionElement`. The same is true for `ForElementImpl`, `ExpressionImpl`, `SpreadElementImpl`, `IfElementImpl`, `MapLiteralEntryImpl`, and `NullAwareElementImpl`. If the mixin `CollectionElementImpl` had not been sealed then the latter would not be true.

The workaround that we usually consider (creating a `class ICollectionElement` declaring all the signatures, declaring `sealed class CollectionElement implements ICollectionElement` and then letting `mixin CollectionElementImpl implements ICollectionElement`) gives rise to a number of inconveniences, and there's no (real ;-) reason why we should not support sealed mixins such that this design can be expressed directly.

**Comment by lrhn** [comment]
That's my point. _Because_ Dart does not have sealed mixins, mixins don't work well in sealed hierarchies.

If you want a sealed hierarchy where a number of separate leaf-subtypes apply a mixin that requires a super-constraint that is in the sealed hierarchy, let's just say the root, then if you do:
```dart
sealed class Base {
  void importantFunction(int x) {}
}
final class Sub1 extends Base with _Mixin {}
final class Sub2 extends Base with _Mixin {}
final class Sub3 extends Base {} // no Mixin
base mixin _Mixin on Base {
  void importantFunction(int x) => super.importantFunction(x + 1);
}
```
then `Sub1`, `Sub2`, `Sub3` will not exhaust `Base`, because `_Mixin` exists as a fourth immediate subtype of `Base` that isn't exhausted by those three.

There are two potential ways to solve that, if the `on` clause is non-negotiable.

1. Allow mixins to be sealed. If `_Mixin` was sealed, and its only immediate subtypes were `Sub1` and `Sub2`, then `Sub1`, `Sub2` and `Sub3` would exhaust `_Mixin` and therefore also `Base`.

2. Not count `_Mixin` as a (significant) immediate subclass of `Base`, so `Sub1`, `Sub2` and `Sub3` exhaust `Base`. The mixin _must_ be `base`, and cannot be implemented even inside the same library. In that case, all instances of that implement `_Mixin` must be an instance of a subclass of one of `Sub1`, `Sub2` or `Sub3`, because _all_ concrete subclasses of `Base` are subclasses of one of those.

Neither is currently possible.

Allowing sealed mixins is probably the easier way to open up this functionality. The "cannot be implemented, ever!" constraint required by the other option is new and confusing.

**Comment by eernstg** [comment]
Sealed mixins were discussed at the language team meeting on May 6, 2026. They are expected to be added to the language soon, unless unexpected difficulties arise. See also https://github.com/dart-lang/sdk/issues/63313.
